File: hello-grpc-python/client/protoClient.py
# encoding: utf-8
"""
gRPC client implementation that demonstrates four gRPC communication patterns.

This module implements a gRPC client that demonstrates:
1. Unary RPC (Talk)
2. Server Streaming RPC (TalkOneAnswerMore)
3. Client Streaming RPC (TalkMoreAnswerOne)
4. Bidirectional Streaming RPC (TalkBidirectional)

This client follows the standardized structure:
1. Configuration constants
2. Logger initialization
3. Connection setup
4. RPC method implementations (unary, server streaming, client streaming, bidirectional)
5. Helper functions
6. Main execution function
7. Cleanup and shutdown
"""

import os
import sys
import logging
import random
import time
import signal

# Add the parent directory to the Python path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from conn import connection, utils, landing_pb2_grpc, landing_pb2, error_mapper

# Configuration constants
RETRY_ATTEMPTS = 3
RETRY_DELAY_SECONDS = 2
ITERATION_COUNT = 3
REQUEST_DELAY_SECONDS = 0.2
SEND_DELAY_SECONDS = 0.002
REQUEST_TIMEOUT_SECONDS = 5
DEFAULT_BATCH_SIZE = 5

# Configure logging
logger = logging.getLogger('grpc-client')
logger.setLevel(logging.INFO)
console = logging.StreamHandler()
console.setLevel(logging.INFO)
formatter = logging.Formatter('[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s',
                              datefmt='%Y-%m-%d %H:%M:%S.%f')
console.setFormatter(formatter)
logger.addHandler(console)

# Common metadata for all requests
DEFAULT_METADATA = (
    ("k1", "v1"),
    ("k2", "v2")
)

# Global flag for graceful shutdown
shutdown_requested = False


def execute_unary_call(stub, request):
    """
    Execute a unary RPC call (single request, single response).
    
    Args:
        stub: The gRPC service stub
        request: The TalkRequest to send
        
    Returns:
        TalkResponse: The response from the server
    """
    request_id = f"unary-{time.time_ns()}"
    logger.info("Sending unary request: data=%s, meta=%s", request.data, request.meta)
    
    start_time = time.time()
    try:
        response = stub.Talk(request=request, metadata=DEFAULT_METADATA, timeout=REQUEST_TIMEOUT_SECONDS)
        duration = time.time() - start_time
        logger.info("Unary call successful in %.3fs", duration)
        return response
    except Exception as e:
        error_mapper.log_error(e, request_id, "Talk")
        raise


def execute_server_streaming_call(stub, request):
    """
    Execute a server streaming RPC call (single request, multiple responses).
    
    Args:
        stub: The gRPC service stub
        request: The TalkRequest to send
    """
    request_id = f"server-stream-{time.time_ns()}"
    logger.info("Starting server streaming with request: data=%s, meta=%s", request.data, request.meta)
    
    start_time = time.time()
    response_count = 0
    
    try:
        response_stream = stub.TalkOneAnswerMore(request=request, metadata=DEFAULT_METADATA, 
                                                  timeout=REQUEST_TIMEOUT_SECONDS)
        
        for response in response_stream:
            if shutdown_requested:
                logger.info("Server streaming cancelled")
                break
            response_count += 1
            logger.info("Received server streaming response #%d:", response_count)
            log_response(response)
        
        duration = time.time() - start_time
        logger.info("Server streaming completed: received %d responses in %.3fs", response_count, duration)
        
    except Exception as e:
        error_mapper.log_error(e, request_id, "TalkOneAnswerMore")
        raise


def execute_client_streaming_call(stub, requests):
    """
    Execute a client streaming RPC call (multiple requests, single response).
    
    Args:
        stub: The gRPC service stub
        requests: List of TalkRequest objects to send
        
    Returns:
        TalkResponse: The response from the server
    """
    request_id = f"client-stream-{time.time_ns()}"
    logger.info("Starting client streaming with %d requests", len(requests))
    
    start_time = time.time()
    request_count = 0
    
    def request_generator():
        nonlocal request_count
        for request in requests:
            if shutdown_requested:
                logger.info("Client streaming cancelled")
                break
            request_count += 1
            logger.info("Sending client streaming request #%d: data=%s, meta=%s", 
                       request_count, request.data, request.meta)
            yield request
            time.sleep(SEND_DELAY_SECONDS)
    
    try:
        response = stub.TalkMoreAnswerOne(request_iterator=request_generator(), 
                                          metadata=DEFAULT_METADATA,
                                          timeout=REQUEST_TIMEOUT_SECONDS)
        duration = time.time() - start_time
        logger.info("Client streaming completed: sent %d requests in %.3fs", request_count, duration)
        return response
        
    except Exception as e:
        error_mapper.log_error(e, request_id, "TalkMoreAnswerOne")
        raise


def execute_bidirectional_streaming_call(stub, requests):
    """
    Execute a bidirectional streaming RPC call (multiple requests, multiple responses).
    
    Args:
        stub: The gRPC service stub
        requests: List of TalkRequest objects to send
    """
    request_id = f"bidirectional-{time.time_ns()}"
    logger.info("Starting bidirectional streaming with %d requests", len(requests))
    
    start_time = time.time()
    request_count = 0
    response_count = 0
    
    def request_generator():
        nonlocal request_count
        for request in requests:
            if shutdown_requested:
                logger.info("Bidirectional streaming cancelled")
                break
            request_count += 1
            logger.info("Sending bidirectional streaming request #%d: data=%s, meta=%s", 
                       request_count, request.data, request.meta)
            yield request
            time.sleep(SEND_DELAY_SECONDS)
    
    try:
        response_stream = stub.TalkBidirectional(request_iterator=request_generator(), 
                                                  metadata=DEFAULT_METADATA,
                                                  timeout=REQUEST_TIMEOUT_SECONDS)
        
        for response in response_stream:
            if shutdown_requested:
                logger.info("Bidirectional streaming cancelled")
                break
            response_count += 1
            logger.info("Received bidirectional streaming response #%d:", response_count)
            log_response(response)
        
        duration = time.time() - start_time
        logger.info("Bidirectional streaming completed in %.3fs", duration)
        
    except Exception as e:
        error_mapper.log_error(e, request_id, "TalkBidirectional")
        raise


def log_response(response):
    """
    Log the details of a response received from the server in a standardized format.
    
    Args:
        response: TalkResponse received from the server
    """
    if response is None:
        logger.warning("Received nil response")
        return
    
    results_count = len(response.results)
    logger.info("Response status: %d, results: %d", response.status, results_count)
    
    for i, result in enumerate(response.results, 1):
        kv = result.kv
        if not kv:
            logger.info("  Result #%d: id=%d, type=%d, kv=nil", i, result.id, result.type)
            continue
        
        meta = kv.get("meta", "")
        result_id = kv.get("id", "")
        idx = kv.get("idx", "")
        data = kv.get("data", "")
        
        logger.info("  Result #%d: id=%d, type=%d, meta=%s, id=%s, idx=%s, data=%s",
                   i, result.id, result.type, meta, result_id, idx, data)


def signal_handler(signum, frame):
    """
    Handle shutdown signals for graceful termination.
    
    Args:
        signum: Signal number
        frame: Current stack frame
    """
    global shutdown_requested
    logger.info("Received shutdown signal, cancelling operations")
    shutdown_requested = True


def run_grpc_calls(stub, delay_seconds, iterations):
    """
    Execute all four gRPC patterns multiple times.
    
    Args:
        stub: The gRPC service stub
        delay_seconds: Delay between iterations in seconds
        iterations: Number of times to run all patterns
        
    Returns:
        bool: True if all calls completed successfully, False otherwise
    """
    for iteration in range(1, iterations + 1):
        if shutdown_requested:
            logger.info("Client execution cancelled")
            return False
        
        logger.info("====== Starting iteration %d/%d ======", iteration, iterations)
        
        try:
            # 1. Unary RPC
            logger.info("----- Executing unary RPC -----")
            unary_request = landing_pb2.TalkRequest(data="0", meta="PYTHON")
            response = execute_unary_call(stub, unary_request)
            log_response(response)
            
            # 2. Server Streaming RPC
            logger.info("----- Executing server streaming RPC -----")
            server_stream_request = landing_pb2.TalkRequest(data="0,1,2", meta="PYTHON")
            execute_server_streaming_call(stub, server_stream_request)
            
            # 3. Client Streaming RPC
            logger.info("----- Executing client streaming RPC -----")
            client_stream_requests = utils.build_link_requests()
            response = execute_client_streaming_call(stub, list(client_stream_requests))
            log_response(response)
            
            # 4. Bidirectional Streaming RPC
            logger.info("----- Executing bidirectional streaming RPC -----")
            bidirectional_requests = utils.build_link_requests()
            execute_bidirectional_streaming_call(stub, list(bidirectional_requests))
            
            # Wait before next iteration, unless it's the last one
            if iteration < iterations:
                logger.info("Waiting %.3fs before next iteration...", delay_seconds)
                time.sleep(delay_seconds)
                
        except Exception as e:
            if shutdown_requested:
                logger.info("Client execution cancelled")
                return False
            logger.error("Error in iteration %d: %s", iteration, e)
            return False
    
    logger.info("All gRPC calls completed successfully")
    return True


def run():
    """
    Main execution function that runs all four gRPC communication patterns.
    """
    # Setup signal handling for graceful shutdown
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    logger.info("Starting gRPC client [version: %s]", utils.get_version())
    
    channel = None
    success = False
    
    # Attempt to establish connection and run all patterns
    for attempt in range(1, RETRY_ATTEMPTS + 1):
        if shutdown_requested:
            logger.info("Client shutting down, aborting retries")
            break
        
        logger.info("Connection attempt %d/%d", attempt, RETRY_ATTEMPTS)
        
        try:
            # Create a gRPC channel and stub
            channel = connection.build_channel()
            stub = landing_pb2_grpc.LandingServiceStub(channel)
            
            # Run all the gRPC patterns
            success = run_grpc_calls(stub, REQUEST_DELAY_SECONDS, ITERATION_COUNT)
            
            if success or shutdown_requested:
                break  # Success or deliberate cancellation, no retry needed
            
        except Exception as e:
            logger.error("Connection attempt %d failed: %s", attempt, e)
            if attempt < RETRY_ATTEMPTS:
                logger.info("Retrying in %ds...", RETRY_DELAY_SECONDS)
                time.sleep(RETRY_DELAY_SECONDS)
            else:
                logger.error("Maximum connection attempts reached, exiting")
        
        finally:
            # Ensure channel is closed for this attempt
            if channel and not success:
                channel.close()
                channel = None
    
    # Final cleanup
    if channel:
        logger.debug("Closing client connection")
        channel.close()
    
    if not success and not shutdown_requested:
        logger.error("Failed to execute all gRPC calls successfully")
        sys.exit(1)
    
    if shutdown_requested:
        logger.info("Client execution was cancelled")
    else:
        logger.info("Client execution completed successfully")


if __name__ == '__main__':
    run()

File: hello-grpc-csharp/HelloClient/ProtoClient.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Threading.Tasks;
using Common;
using Grpc.Core;
using Hello;
using log4net;
using log4net.Config;

namespace HelloClient
{
    /// <summary>
    /// gRPC client implementation demonstrating all four RPC patterns.
    /// 
    /// This client follows the standardized structure:
    /// 1. Configuration constants
    /// 2. Logger initialization
    /// 3. Connection setup
    /// 4. RPC method implementations (unary, server streaming, client streaming, bidirectional)
    /// 5. Helper functions
    /// 6. Main execution function
    /// 7. Cleanup and shutdown
    /// </summary>
    public class ProtoClient
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(ProtoClient));
        
        // Configuration constants
        private const int RetryAttempts = 3;
        private const int RetryDelayMs = 2000;
        private const int IterationCount = 3;
        private const int RequestDelayMs = 200;
        private const int SendDelayMs = 2;
        private const int RequestTimeoutSeconds = 5;
        
        private readonly LandingService.LandingServiceClient _client;

        private ProtoClient(LandingService.LandingServiceClient client)
        {
            this._client = client;
        }

        public static void Main()
        {
            // Configure log4net
            ConfigureLogging();
            
            Log.Info($"Starting gRPC client [version: {Utils.GetVersion()}]");

            ProtoClient? protoClient = null;
            bool success = false;

            // Attempt to establish connection and run all patterns
            for (int attempt = 1; attempt <= RetryAttempts; attempt++)
            {
                Log.Info($"Connection attempt {attempt}/{RetryAttempts}");

                try
                {
                    var channel = Connection.GetChannel();
                    protoClient = new ProtoClient(new LandingService.LandingServiceClient(channel));

                    // Run all the gRPC patterns
                    RunGrpcCalls(protoClient, RequestDelayMs, IterationCount).Wait();
                    success = true;
                    
                    // Cleanup
                    channel.ShutdownAsync().Wait();
                    break; // Success, no retry needed
                }
                catch (Exception e)
                {
                    Log.Error($"Connection attempt {attempt} failed: {e.Message}");
                    if (attempt < RetryAttempts)
                    {
                        Log.Info($"Retrying in {RetryDelayMs}ms...");
                        Task.Delay(RetryDelayMs).Wait();
                    }
                    else
                    {
                        Log.Error("Maximum connection attempts reached, exiting");
                    }
                }
            }

            if (!success)
            {
                Log.Error("Failed to execute all gRPC calls successfully");
                Environment.Exit(1);
            }

            Log.Info("Client execution completed successfully");
        }

        /// <summary>
        /// Configures log4net logging framework
        /// </summary>
        private static void ConfigureLogging()
        {
            string baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
            string configPath = Path.Combine(baseDirectory, "log4net.config");
            
            if (!File.Exists(configPath))
            {
                string currentDirectory = Directory.GetCurrentDirectory();
                string projectDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) ?? "";
                
                if (File.Exists("log4net.config"))
                    configPath = "log4net.config";
                else if (File.Exists(Path.Combine(currentDirectory, "log4net.config")))
                    configPath = Path.Combine(currentDirectory, "log4net.config");
                else if (File.Exists(Path.Combine(projectDir, "log4net.config")))
                    configPath = Path.Combine(projectDir, "log4net.config");
            }
            
            var logRepository = LogManager.GetRepository(Assembly.GetEntryAssembly()!);
            XmlConfigurator.Configure(logRepository, new FileInfo(configPath));
            
            // Ensure console output uses auto-flush
            Console.SetOut(new StreamWriter(Console.OpenStandardOutput()) { AutoFlush = true });
        }

        /// <summary>
        /// Executes all four gRPC patterns multiple times
        /// </summary>
        /// <param name="client">The client instance to use</param>
        /// <param name="delayMs">Delay between iterations in milliseconds</param>
        /// <param name="iterations">Number of times to run all patterns</param>
        private static async Task RunGrpcCalls(ProtoClient client, int delayMs, int iterations)
        {
            for (int iteration = 1; iteration <= iterations; iteration++)
            {
                Log.Info($"====== Starting iteration {iteration}/{iterations} ======");

                // 1. Unary RPC
                Log.Info("----- Executing unary RPC -----");
                var unaryRequest = new TalkRequest { Data = "0", Meta = "C#" };
                var response = client.ExecuteUnaryCall(unaryRequest);
                LogResponse(response);

                // 2. Server streaming RPC
                Log.Info("----- Executing server streaming RPC -----");
                var serverStreamRequest = new TalkRequest { Data = "0,1,2", Meta = "C#" };
                await client.ExecuteServerStreamingCall(serverStreamRequest);

                // 3. Client streaming RPC
                Log.Info("----- Executing client streaming RPC -----");
                var clientStreamResponse = await client.ExecuteClientStreamingCall(Utils.BuildLinkRequests());
                LogResponse(clientStreamResponse);

                // 4. Bidirectional streaming RPC
                Log.Info("----- Executing bidirectional streaming RPC -----");
                await client.ExecuteBidirectionalStreamingCall(Utils.BuildLinkRequests());

                // Wait before next iteration, unless it's the last one
                if (iteration < iterations)
                {
                    Log.Info($"Waiting {delayMs}ms before next iteration...");
                    await Task.Delay(delayMs);
                }
            }

            Log.Info("All gRPC calls completed successfully");
        }

        /// <summary>
        /// Demonstrates the unary RPC pattern (single request, single response)
        /// </summary>
        /// <param name="request">The request to send</param>
        /// <returns>The response from the server</returns>
        public TalkResponse ExecuteUnaryCall(TalkRequest request)
        {
            string requestId = $"unary-{DateTime.Now.Ticks}";
            Log.Info($"Sending unary request: data={request.Data}, meta={request.Meta}");

            var stopwatch = Stopwatch.StartNew();
            try
            {
                var response = _client.Talk(request, BuildHeaders());
                stopwatch.Stop();
                Log.Info($"Unary call successful in {stopwatch.ElapsedMilliseconds}ms");
                return response;
            }
            catch (RpcException e)
            {
                ErrorMapper.LogError(e, requestId, "Talk");
                throw;
            }
        }

        /// <summary>
        /// Demonstrates the server streaming RPC pattern (single request, multiple responses)
        /// </summary>
        /// <param name="request">The request to send</param>
        public async Task ExecuteServerStreamingCall(TalkRequest request)
        {
            string requestId = $"server-stream-{DateTime.Now.Ticks}";
            Log.Info($"Starting server streaming with request: data={request.Data}, meta={request.Meta}");

            var stopwatch = Stopwatch.StartNew();
            int responseCount = 0;

            try
            {
                using var call = _client.TalkOneAnswerMore(request, BuildHeaders());
                var responseStream = call.ResponseStream;
                
                while (await responseStream.MoveNext())
                {
                    responseCount++;
                    var response = responseStream.Current;
                    Log.Info($"Received server streaming response #{responseCount}:");
                    LogResponse(response);
                }

                stopwatch.Stop();
                Log.Info($"Server streaming completed: received {responseCount} responses in {stopwatch.ElapsedMilliseconds}ms");
            }
            catch (RpcException e)
            {
                ErrorMapper.LogError(e, requestId, "TalkOneAnswerMore");
                throw;
            }
        }

        /// <summary>
        /// Demonstrates the client streaming RPC pattern (multiple requests, single response)
        /// </summary>
        /// <param name="requests">The list of requests to send</param>
        /// <returns>The response from the server</returns>
        public async Task<TalkResponse> ExecuteClientStreamingCall(System.Collections.Generic.LinkedList<TalkRequest> requests)
        {
            string requestId = $"client-stream-{DateTime.Now.Ticks}";
            Log.Info($"Starting client streaming with {requests.Count} requests");

            var stopwatch = Stopwatch.StartNew();
            int requestCount = 0;

            try
            {
                using var call = _client.TalkMoreAnswerOne(BuildHeaders());
                
                foreach (var request in requests)
                {
                    requestCount++;
                    Log.Info($"Sending client streaming request #{requestCount}: data={request.Data}, meta={request.Meta}");
                    await call.RequestStream.WriteAsync(request);
                    await Task.Delay(SendDelayMs);
                }
                
                await call.RequestStream.CompleteAsync();
                var response = await call.ResponseAsync;
                
                stopwatch.Stop();
                Log.Info($"Client streaming completed: sent {requestCount} requests in {stopwatch.ElapsedMilliseconds}ms");
                
                return response;
            }
            catch (RpcException e)
            {
                ErrorMapper.LogError(e, requestId, "TalkMoreAnswerOne");
                throw;
            }
        }

        /// <summary>
        /// Demonstrates the bidirectional streaming RPC pattern (multiple requests, multiple responses)
        /// </summary>
        /// <param name="requests">The list of requests to send</param>
        public async Task ExecuteBidirectionalStreamingCall(System.Collections.Generic.LinkedList<TalkRequest> requests)
        {
            string requestId = $"bidirectional-{DateTime.Now.Ticks}";
            Log.Info($"Starting bidirectional streaming with {requests.Count} requests");

            var stopwatch = Stopwatch.StartNew();
            int requestCount = 0;
            int responseCount = 0;

            try
            {
                using var call = _client.TalkBidirectional(BuildHeaders());
                
                // Start receiving responses in a separate task
                var responseReaderTask = Task.Run(async () =>
                {
                    while (await call.ResponseStream.MoveNext())
                    {
                        responseCount++;
                        var response = call.ResponseStream.Current;
                        Log.Info($"Received bidirectional streaming response #{responseCount}:");
                        LogResponse(response);
                    }
                });
                
                // Send all requests
                foreach (var request in requests)
                {
                    requestCount++;
                    Log.Info($"Sending bidirectional streaming request #{requestCount}: data={request.Data}, meta={request.Meta}");
                    await call.RequestStream.WriteAsync(request);
                    await Task.Delay(SendDelayMs);
                }
                
                // Close sending side
                Log.Info("Closing send side of bidirectional stream");
                await call.RequestStream.CompleteAsync();
                
                // Wait for receiving side to complete
                await responseReaderTask;
                
                stopwatch.Stop();
                Log.Info($"Bidirectional streaming completed in {stopwatch.ElapsedMilliseconds}ms");
            }
            catch (RpcException e)
            {
                ErrorMapper.LogError(e, requestId, "TalkBidirectional");
                throw;
            }
        }

        /// <summary>
        /// Builds standard metadata headers for requests
        /// </summary>
        /// <returns>Metadata with standard headers</returns>
        private static Metadata BuildHeaders()
        {
            var headers = new Metadata
            {
                new Metadata.Entry("k1", "v1"),
                new Metadata.Entry("k2", "v2")
            };
            return headers;
        }

        /// <summary>
        /// Logs the response in a standardized format
        /// </summary>
        /// <param name="response">The response to log</param>
        private static void LogResponse(TalkResponse response)
        {
            if (response == null)
            {
                Log.Warn("Received nil response");
                return;
            }

            int resultsCount = response.Results.Count;
            Log.Info($"Response status: {response.Status}, results: {resultsCount}");

            for (int i = 0; i < resultsCount; i++)
            {
                var result = response.Results[i];
                var kv = result.Kv;

                string meta = kv.ContainsKey("meta") ? kv["meta"] : "";
                string id = kv.ContainsKey("id") ? kv["id"] : "";
                string idx = kv.ContainsKey("idx") ? kv["idx"] : "";
                string data = kv.ContainsKey("data") ? kv["data"] : "";

                Log.Info($"  Result #{i + 1}: id={result.Id}, type={result.Type}, meta={meta}, id={id}, idx={idx}, data={data}");
            }
        }
    }
}